### PR TITLE
Normalize

### PR DESCRIPTION
**Pull request by seleanridley**

### CODE REVIEW

File: cpp/include/cpu_operations.h [hunk 1]
```diff
@@ -162,24 +162,28 @@ static Vector<T> Norm(const Matrix<T> &a,
...
      }
      return norm;
-     } else {
+    } else if (axis == 1) {
+     Vector<T> norm(num_rows);
      for (int i = 0; i < num_rows; i++) {
       for (int j = 0; j < num_cols; j++)
          nval += pow(a(i, j), p);
        norm(i) = pow(nval, (1.0/p));
        nval = 0;
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: what does nval stands for? Is it norm_value?

File: cpp/include/cpu_operations.h [hunk 2]
```diff
@@ -254,27 +258,21 @@ static Vector<T> Norm(const Matrix<T> &a,
...
-     for (int j = 0; j < num_cols; j++) {
-      for (int i = 0; i < num_rows; i++) {
-       float norm = a.col(j).norm();
-       b(i, j) = a(i, j)/norm;
-       }
-     }
-    return b;
+     b = a.transpose().array().colwise() / Norm(a, p, axis).array();
+     return b.transpose();
+     } else if (axis == 1) {
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: the closing curly brace should be indented the same as the if statement

File: cpp/include/cpu_operations.h [hunk 3]
```diff
@@ -254,27 +258,21 @@ static Vector<T> Norm(const Matrix<T> &a,
...
+        b = a.array().colwise() / Norm(a, p, axis).array();
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: b here should have the same indentation as the b in the if statement block

File: cpp/include/cpu_operations.h [hunk 4]
```diff
@@ -254,27 +258,21 @@ static Vector<T> Norm(const Matrix<T> &a,
...
      } else {
-      for (int i = 0; i < num_rows; i++) {
-       for (int j = 0; j < num_cols; j++) {
-        float norm = a.row(i).norm();
-        b(i, j) = a(i, j)/norm;
-       }
-     }
-     return b;
-     }
+        std::cerr << "Axis must be zero or one!";
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: same indentation for std::cerr as the b in the if block

File: cpp/include/cpu_operations.h [hunk 5]
```diff
@@ -254,27 +258,21 @@ static Vector<T> Norm(const Matrix<T> &a,
...
-       float norm = a.col(j).norm();
-       b(i, j) = a(i, j)/norm;
-       }
-     }
-    return b;
+     b = a.transpose().array().colwise() / Norm(a, p, axis).array();
+     return b.transpose();
+     } else if (axis == 1) {
+        b = a.array().colwise() / Norm(a, p, axis).array();
+        return b;
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: The indentation should be 2 spaces, it seems you have a 4 spaces indentation.

File: cpp/include/cpu_operations.h (hunk 1 above)
Comment (seleanridley) [inline cpp/include/cpu_operations.h]: Not necessarily. I couldn't think of a good name for that variable but I'll try to change it to something clearer.

File: cpp/include/cpu_operations.h (hunk 5 above)
Comment (seleanridley) [inline cpp/include/cpu_operations.h]: Yes, sorry. That was before I changed the indentation in my vim config.

File: cpp/include/cpu_operations.h [hunk 6]
```diff
@@ -266,13 +273,13 @@ static Vector<T> Norm(const Matrix<T> &a,
...
-        std::cerr << "Axis must be zero or one!";
-        exit(1);
-        }
+    } else if (axis == 1) {
+     b = a.array().colwise() / Norm(a, p, axis).array();
+     return b;
+    } else {
+     std::cerr << "Axis must be zero or one!";
+     exit(1);
+      }
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: This should be aligned with "static" in "static Matrix<T> Normalize"

File: cpp/include/cpu_operations.h [hunk 7]
```diff
@@ -266,13 +273,13 @@ static Vector<T> Norm(const Matrix<T> &a,
...
 }
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: This should be aligned with "static" in "static Matrix<T> Normalize"

File: cpp/include/cpu_operations.h (hunk 6 above)
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: This should be aligned with the } before else

File: cpp/include/cpu_operations.h [hunk 8]
```diff
@@ -266,13 +273,13 @@ static Vector<T> Norm(const Matrix<T> &a,
...
-     } else {
-        std::cerr << "Axis must be zero or one!";
-        exit(1);
-        }
+    } else if (axis == 1) {
+     b = a.array().colwise() / Norm(a, p, axis).array();
+     return b;
+    } else {
+     std::cerr << "Axis must be zero or one!";
+     exit(1);
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h:397]: Could you also add some doxygen comments to Norm and Normalize?

File: cpp/include/cpu_operations.h [hunk 9]
```diff
@@ -156,7 +156,28 @@ class CpuOperations {
...
+/// const Matrix <T> &a
+/// \param b
+/// \const int &p 
+/// \param c
+/// \const int &axis
+///
+/// \return 
+/// Vector <T>
+/// \sa 
+/// \ref Frobenius Norm : Frobenius Norm is similar to Norm where
```
Comment (yiskylee) [inline cpp/include/cpu_operations.h]: Frobenius norm actually is different: it essentially treat a matrix as a 1d vector and returns the norm of that vector. Could you change your comment?

Comment (yiskylee) [inline cpp/include/cpu_operations.h]: Also here the \ref doesn't give me link to Frobenius Norm. Could you make a link? Try 
\ref FrobeniusNorm or something. You could discuss with Leslie on how to make this work.

**Comment by yiskylee** [comment]
See my line comments.
Also, could you merge from upstream develop before I could merge your pull request?

**Pull request merged**